### PR TITLE
Interactivity API: Skip instead of bail out if HTML contains SVG or MATH

### PR DESCRIPTION
**Pull request by cbravobernal**
When there is a SVG icon in the navigation menu. The Interactivity API is not server-side processing the directives. This is causing a flash effect when the runtime process the JavaScript part.

Trac ticket: https://core.trac.wordpress.org/ticket/60517

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Review by westonruter — APPROVED** [review]
LGTM!

File: src/wp-content/fonts/OpenSans-Regular.ttf
Comment (westonruter) [inline src/wp-content/fonts/OpenSans-Regular.ttf:1]: Oh wait, why is this TTF file being included in this PR?

**Review by westonruter — CHANGES_REQUESTED** [review]

File: src/wp-content/fonts/OpenSans-Regular.ttf
Comment (cbravobernal) [reply src/wp-content/fonts/OpenSans-Regular.ttf:1]: That's weird, sorry 🙇

**Review by westonruter — APPROVED** [review]

**Review by dmsnell — COMMENTED** [review]
This has been something that @sirreal and I have been working with lately.

WordPress/wordpress-develop#6006

it's complicated, and I might recommend creating something marginally more robust in the meantime, because the rules for the self-closing flag change within SVG and MathML.

that is, non-void elements which contain the self-closing flag are treated as empty elements.

there are escapes from SVG and MathML that I think we can practically overlook for now because the logic for detecting them is complicated, but a new method `$p->skip_foreign_content()` might serve us better.

this method would be _like_ `next_balanced_tag()` except it will only pay attention to the current tag.

```php
private function skip_foreign_content() {
	$tag_name = $this->get_tag_name();
	$depth    = 1;

	while ( $depth > && $this->next_tag( $tag_name ) ) {
		if ( $this->has_self_closing_flag() ) {
			continue;
		}

		$depth += $this->is_tag_closer() ? -1 : 1;
	}

	return 0 === $depth;
}
```

**Comment by cbravobernal** [comment]
> This has been something that @sirreal and I have been working with lately.
> 
> #6006
> 
> it's complicated, and I might recommend creating something marginally more robust in the meantime, because the rules for the self-closing flag change within SVG and MathML.
> 
> that is, non-void elements which contain the self-closing flag are treated as empty elements.
> 
> there are escapes from SVG and MathML that I think we can practically overlook for now because the logic for detecting them is complicated, but a new method `$p->skip_foreign_content()` might serve us better.
> 
> this method would be _like_ `next_balanced_tag()` except it will only pay attention to the current tag.
> 
> ```
> private function skip_foreign_content() {
> 	$tag_name = $this->get_tag_name();
> 	$depth    = 1;
> 
> 	while ( $depth > && $this->next_tag( $tag_name ) ) {
> 		if ( $this->has_self_closing_flag() ) {
> 			continue;
> 		}
> 
> 		$depth += $this->is_tag_closer() ? -1 : 1;
> 	}
> 
> 	return 0 === $depth;
> }
> ```

I created the function as suggested, with small tweaks to make it work as expected. I also added some tests in a pair programming session with @michalczaplinski

I would feel much more confident with a second review, thanks @dmsnell!

File: src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php [hunk 1]
```diff
@@ -180,6 +180,39 @@ private function get_balanced_tag_bookmarks() {
...
+		while ( $depth > 0 && $this->next_tag(
+			array(
+				'tag_name'    => $tag_name,
+				'tag_closers' => 'visit',
+			)
+		) ) {
+			if ( $this->has_self_closing_flag() ) {
+				continue;
+			}
+			$depth += $this->is_tag_closer() ? -1 : 1;
```
Comment (westonruter) [inline src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php:210]: @dmsnell What about a `title` tag that appears inside of an SVG? Since it's in XML parsing mode, would it still be processed as RCDATA here?

**Comment by cbravobernal** [comment]
Sorry @westonruter for re-requesting your review again, but the code has change quite a bit 😅

File: src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php (hunk 1 above)
Comment (dmsnell) [reply src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php:210]: @westonruter this is a great question, and you're right. it's not RCDATA when found inside SVG. this is all the more reason to focus only on SVG and MathML tags and ignore the others.

in https://github.com/WordPress/wordpress-develop/pull/6006 what I'm looking to do is have the HTML Processor tell the Tag Processor that it's inside foreign content, and while that flag is active, we'll disable the special handling of tokens like TITLE and the others.

it's complicated and we explored trying to put the foreign-content handling code inside the Tag Processor, but that ended up demanding that we re-implement much of the HTML Processor's functionality there, so it will probably be the case that you need the HTML Processor to adequately parse inside SVG and MathML.

in the meantime, if one overlooks the HTML elements which break out into the nearest "integration point," then I think that skipping from SVG tag to SVG tag will be reliable enough for now.

Comment (dmsnell) [reply src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php:210]: aha I misread your statement @westonruter concerning `TITLE`. in this case it shouldn't matter that we're checking `is_tag_closer()` because we're not visiting the `TITLE` tag - we're skipping straight to whichever tag (`SVG` or `MATH`) opened the foreign content.

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php [hunk 2]
```diff
@@ -508,50 +508,130 @@ public function test_process_directives_doesnt_change_html_if_contains_unbalance
...
+	public function test_process_directives_does_not_change_inner_html_in_svgs() {
+		$this->interactivity->state(
+			'myPlugin',
+			array(
+				'id' => 'some-id',
+			)
+		);
+		$html           = '
+			<header>
+				<svg height="100">
```
Comment (westonruter) [inline tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php:562]: Let's add a `title` tag here just to make sure it works as expected.

```suggestion
				<svg height="100">
					<title>Red Circle</title>
```

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php [hunk 3]
```diff
@@ -508,50 +508,130 @@ public function test_process_directives_doesnt_change_html_if_contains_unbalance
...
+					<circle cx="50" cy="50" r="40" stroke="black" stroke-width="3" fill="red" />
+					<div data-wp-bind--id="myPlugin::state.id"></div>
```
Comment (westonruter) [inline tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: Why not an SVG element?

```suggestion
					<g data-wp-bind--id="myPlugin::state.id" />
```

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php [hunk 4]
```diff
@@ -508,50 +508,130 @@ public function test_process_directives_doesnt_change_html_if_contains_unbalance
...
+		$this->interactivity->state(
+			'myPlugin',
+			array(
+				'id' => 'some-id',
+			)
+		);
+		$html           = '
+			<header>
+				<math data-wp-bind--id="myPlugin::state.math">
+				<div data-wp-bind--id="myPlugin::state.id"></div>
```
Comment (westonruter) [inline tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: Probably better to use actual MathML, right?

```suggestion
					<mrow data-wp-bind--id="myPlugin::state.id" />
```

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPIDirectivesProcessor.php [hunk 5]
```diff
@@ -778,4 +778,51 @@ public function test_next_balanced_tag_closer_tag_on_closing_tag() {
...
+	 *
+	 * @ticket 60517
+	 *
+	 * @covers ::skip_foreign_content
+	 */
+	public function test_skip_foreign_content() {
+		$content = '<div><span>Not closed</div>';
+		$p       = new WP_Interactivity_API_Directives_Processor( $content );
+		$p->next_tag();
+		$this->assertTrue( $p->skip_foreign_content() );
```
Comment (westonruter) [inline tests/phpunit/tests/interactivity-api/wpInteractivityAPIDirectivesProcessor.php]: There's actually no foreign content here. Should a better name be chosen than `skip_foreign_content()` like `jump_to_tag_closer()`?

**Review by westonruter — CHANGES_REQUESTED** [review]

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php (hunk 3 above)
Comment (dmsnell) [reply tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: this is a good example of a bad example anyway 🙃 as the `div` here would close the open `SVG` element 🤦‍♂️

the full set of HTML tags which break out to the nearest integration point are:

> "b", "big", "blockquote", "body", "br", "center", "code", "dd", "div", "dl", "dt", "em", "embed", "h1", "h2", "h3", "h4", "h5", "h6", "head", "hr", "i", "img", "li", "listing", "menu", "meta", "nobr", "ol", "p", "pre", "ruby", "s", "small", "span", "strong", "strike", "sub", "sup", "table", "tt", "u", "ul", "var"

and some `FONT` start tags containing specific attributes. also some end tags.

https://html.spec.whatwg.org/#parsing-main-inforeign

the "integration point" thing is a bit more complicated than it appears at a glance. there can be a stack of open elements that zigzags between HTML and foreign content, and some SVG/MathML tags themselves create new integration points, so a `DIV` inside one of those regions won't close out the `SVG` - it will only close out everything up until that element.

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php (hunk 4 above)
Comment (cbravobernal) [reply tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: > The idea was to use an element that we know the server directive processing would process. As far as I know, it should skip SVG tags and its inner accepted elements by default.

Similar here but with MathML

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php (hunk 3 above)
Comment (cbravobernal) [reply tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: The idea was to use an element that we know the server directive processing would process. As far as I know, it should skip SVG tags and its inner accepted elements by default.

**Comment by cbravobernal** [comment]
I addressed the requested changes in https://github.com/WordPress/wordpress-develop/pull/6091/commits/23f25c1cc8cdf766510e9f64f2322fef91ed07a3

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php (hunk 3 above)
Comment (dmsnell) [reply tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: @c4rl0sbr4v0 my point was mostly that in an HTML document, the `DIV` in this example closes out the `SVG`, and so the `skip_foreign_content()` function would fail here.

if you wanted to use an element that's neutral on this, it would be fine to use something like `<aside>` or `<data>` or `<abbr>`, as these and any other elements not listed above do not affect the foreign content stack.

Comment (cbravobernal) [reply tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: If they are edge cases, maybe is better to just use what should be inside a SVG tag.

> if you wanted to use an element that's neutral on this, it would be fine to use something like <aside> or <data> or <abbr>, as these and any other elements not listed above do not affect the foreign content stack.

You are becoming a huge expert in HTML processing @dmsnell 😃

File: src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php [hunk 6]
```diff
@@ -180,6 +180,39 @@ private function get_balanced_tag_bookmarks() {
...
+	 * This function is intended to skip processing SVG and MathML inner content
+	 * instead of bailing out the whole processing.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @access private
+	 *
+	 * @return bool Whether the foreign content was successfully skipped.
+	 */
+	public function jump_to_tag_closer(): bool {
```
Comment (westonruter) [inline src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php]: Just realized that `skip` is the terminology used in the tag processor and also in the description so it would probably be better here.

```suggestion
	public function skip_to_tag_closer(): bool {
```

File: tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php [hunk 7]
```diff
@@ -508,50 +508,131 @@ public function test_process_directives_doesnt_change_html_if_contains_unbalance
...
+		$this->interactivity->state(
+			'myPlugin',
+			array(
+				'id' => 'some-id',
+			)
+		);
+		$html           = '
+			<header>
+				<math data-wp-bind--id="myPlugin::state.math">
+				<mrow data-wp-bind--id="myPlugin::state.id" />
```
Comment (westonruter) [inline tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php]: Indentation nit.

```suggestion
					<mrow data-wp-bind--id="myPlugin::state.id" />
```

**Review by westonruter — APPROVED** [review]
Just a couple nits left, so I'm pre-approving.

File: src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php (hunk 6 above)
Comment (cbravobernal) [reply src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php]: That's true.

**Comment by cbravobernal** [comment]
> Just a couple nits left, so I'm pre-approving.

Done in https://github.com/WordPress/wordpress-develop/pull/6091/commits/8ea4a695c99b77813fc8fa1b37dd9c94bafb27b3

Ready to commit then!

**Comment by swissspidy** [comment]
@dmsnell Is this one good to go in your opinion as well?

**Comment by cbravobernal** [comment]
> @dmsnell Is this one good to go in your opinion as well?

I'm afraid Dennis is AFK until the 26th of February. Not sure if he will answer on time 😅

**Comment by swissspidy** [comment]
Alright. Committed in https://core.trac.wordpress.org/changeset/57649. We can always commit follow-ups if needed.

**Comment by cbravobernal** [comment]
@sirreal has a lot of experience with HTML API, he could take a look at it too

**Comment by sirreal** [comment]
👍 The changes as committed here look good to me.

**Comment by dmsnell** [comment]
All good from my end too. Of course we could make this more robust, but doing so inches us towards the HTML Processor and isn't worth recreating. I think we're going to start exploring using the HTML Processor for the Interactivity API, so in time this will be a non-issue.
